### DISCUSSION TITLE
Converting SARGE to Tskit?

### DISCUSSION THREAD

**Discussion by swamidass** [tskit-dev/tskit#1879] [Q&A] [answered]
@nkschaefer just published an ARG of humans, denisovans, and neanderthals using SARGE.

They did not use tskit, but used SARGE. 

https://github.com/nkschaefer/sarge

Is there any way to convert a SARGE arg into a tskit ARG?

**Comment by jeromekelleher** [comment, marked as answer]
Not that I know of @swamidass, I don't think anyone has looked at this. I guess ideally we'd like SARGE to support converting to tskit rather than us trying to reverse engineer things. It would probably be a pretty straightforward utility program using our C API.

Maybe we should open an issue on their repo to ask for this feature?

**Reply by benjeffery** [reply]
Seems they wish to interoperate as they have things like https://github.com/nkschaefer/sarge/blob/master/utilities/run_tsinfer.sh

**Reply by swamidass** [reply]
That's my impression too @benjeffery

**Comment by swamidass** [comment]
Putting an issue in with him is a good idea. I suspect open lines of communication between @nkschaefer and the TSKit team would be ideal. It does seem that it is possible to convert SARGE files to a sequence of newick trees. From Nick regarding the SARGE codebase:

> One program, for example, is called trees2newick and will dump each chromosome, site position, and tree (in newick format), as tab separated output, one tree per line.  

The problem, however, is that I'm not sure if this will produce a tree sequence that meets all your constraints. Also, it would seem that the distribution of mutations on the trees may be lost.

---


As a reference point, SARGE was used to calculate the ARG in this paper:

https://www.science.org/doi/10.1126/sciadv.abc0776

I am honestly curious to whether or not tsinfer produces a comparable ARG or not. It seems that using ancient DNA makes the analysis more complex though, and I'm not sure if or how tsinfer handles the higher error rates, etc. If it isn't impossible, perhaps they'd be willing to share the input data they used, to run tsinfer on too.

**Reply by jeromekelleher** [reply]
It would be great to know, I agree @swamidass! As I say, it would be really excellent if they could provide a method to convert directly from their format to tskit. Since we provide a documented C API that should be very feasible, even straightforward.

Maybe you could open an issue on the SARGE repo with a feature request, so folks know there's a general desire for this feature? I'm very happy to weigh in with technical help.

**Comment by swamidass** [comment]
It also looks like he has code to convert tskit trees into SARGE. I wonder if that gives a good enough hook to work with...

https://github.com/nkschaefer/sarge/blob/master/utilities/sarge_tsinfer.py
